be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp

```diff
@@ -0,0 +1,82 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#include "aggregate_blocking_sink_operator.h"
+
+namespace starrocks::pipeline {
+
+Status AggregateBlockingSinkOperator::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(Operator::prepare(state));
+    // _aggregator is shared by sink operator and source operator
+    // we must only prepare it at sink operator
+    return _aggregator->prepare(state, state->obj_pool(), get_memtracker(), get_runtime_profile());
+}
+
+bool AggregateBlockingSinkOperator::is_finished() const {
+    return _is_finished;
+}
+
+void AggregateBlockingSinkOperator::finish(RuntimeState* state) {
+    if (_is_finished) {
+        return;
+    }
+    _is_finished = true;
+
+    if (!_aggregator->is_none_group_by_exprs()) {
+        COUNTER_SET(_aggregator->hash_table_size(), (int64_t)_aggregator->hash_map_variant().size());
+        // If hash map is empty, we don't need to return value
+        if (_aggregator->hash_map_variant().size() == 0) {
+            _aggregator->set_ht_done();
+        }
+
+        if (false) {
+        }
+#define HASH_MAP_METHOD(NAME)                                                                                         \
+    else if (_aggregator->hash_map_variant().type == vectorized::HashMapVariant::Type::NAME) _aggregator->it_hash() = \
+            _aggregator->hash_map_variant().NAME->hash_map.begin();
+        APPLY_FOR_VARIANT_ALL(HASH_MAP_METHOD)
+#undef HASH_MAP_METHOD
+    } else if (_aggregator->is_none_group_by_exprs()) {
+        // for aggregate no group by, if _num_input_rows is 0,
+        // In update phase, we directly return empty chunk.
+        // In merge phase, we will handle it.
+        if (_aggregator->num_input_rows() == 0 && !_aggregator->is_needs_finalize()) {
+            _aggregator->set_ht_done();
+        }
+    }
+    COUNTER_SET(_aggregator->input_row_count(), _aggregator->num_input_rows());
+
+    _aggregator->sink_complete();
```

Comment:
what's the difference between _is_sink_complete and _is_ht_done?

Reply:
sink_complete means: no further chunk will push into sink operator

Reply:
adjust is_ht_done to is_ht_eos